compiler/src/iree/compiler/Codegen/SPIRV/test/pipeline_reduction_subgroup.mlir

```diff
@@ -149,3 +149,142 @@ hal.executable private @subgroup_reduce {
 }
 
 // CHECK-NOT: spirv.GroupNonUniformShuffleXor
+
+// -----
+
+#pipeline_layout = #hal.pipeline.layout<push_constants = 2, sets = [
+  #hal.descriptor_set.layout<0, bindings = [
+    #hal.descriptor_set.binding<0, storage_buffer, ReadOnly>,
+    #hal.descriptor_set.binding<1, storage_buffer>
+  ]>
+]>
+
+hal.executable private @dynamic_softmax {
+  hal.executable.variant public @vulkan_spirv_fb target(<"vulkan", "vulkan-spirv-fb", {
+      spirv.target_env = #spirv.target_env<#spirv.vce<v1.6,
+        [Shader, Float16, StorageBuffer16BitAccess, StorageUniform16, GroupNonUniformShuffle],
+        [SPV_KHR_16bit_storage]>, api=Vulkan, Unknown:DiscreteGPU, #spirv.resource_limits<
+          max_compute_shared_memory_size = 65536,
+          max_compute_workgroup_invocations = 1024,
+          max_compute_workgroup_size = [1024, 1024, 1024],
+          subgroup_size = 64>>
+    }>) {
+    hal.executable.export public @dynamic_softmax ordinal(0) layout(#pipeline_layout) {
+    ^bb0(%arg0: !hal.device, %arg1: index):
+      %x, %y, %z = flow.dispatch.workgroup_count_from_slice %arg1
+      hal.return %x, %y, %z : index, index, index
+    }
+    builtin.module {
+      func.func @dynamic_softmax() {
+        %c32_i64 = arith.constant 32 : i64
+        %c0 = arith.constant 0 : index
+        %0 = hal.interface.constant.load[0] : i32
+        %1 = hal.interface.constant.load[1] : i32 
+        %2 = arith.extui %0 : i32 to i64
+        %3 = arith.extui %1 : i32 to i64
+        %4 = arith.shli %3, %c32_i64 : i64 
+        %5 = arith.ori %2, %4 : i64
+        %6 = arith.index_castui %5 : i64 to index
+        %7 = flow.dispatch.workload.ordinal %6, 0 : index
+        %8 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<32x?xf16>>{%7}
+        %9 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<32x?xf16>>{%7}
+        %10 = flow.dispatch.tensor.load %8, offsets = [0, 0], sizes = [32, %7], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<32x?xf16>>{%7} -> tensor<32x?xf16>
+        %11 = tensor.empty(%7) : tensor<32x?xf16> 
+        %12 = linalg.softmax dimension(1) ins(%10 : tensor<32x?xf16>) outs(%11 : tensor<32x?xf16>) -> tensor<32x?xf16>
```

Comment:
I'm expecting we may want to make changes to this pipeline in the relative near future, but would like to keep this test simple to show that we get masking + distribution right. Would it be better to decompose the softmax within the test and skip lowering configuration for it? (just setting the lowering configurations explicitly on the op).

Reply:
Depends on how comfortable you are with the test. 

Typically we need to use `iree-codegen-linalg-to-spirv-pipeline` when the full conversion flow, including the memref and SPIR-V lowering part, is interesting. But most of the time lowering memref and converting to SPIR-V is mechanical; so `iree-spirv-lower-executable-target-pass` (see those `SPIRV/test/lowering_*.mlir` tests) would do--it would already guarantee the big picture stuff. I think this might apply here. We do have some modifications to SPIRVVectorizeLoadStore, but that's mostly mechanical and guaranteed by their own tests. So checking at the memref and vector level should be good.

Embedding lowering_config is good when we want to make the tests detach from the lowering configure deduction logic, which already have their own tests (those `SPIRV/test/config_*.mlir` tests). We don't have very sophisticated config deduction logic for dynamic-dim reduction; so I suppose it's fine so I'm okay both.

Reply:
Oh nice, checking at the memref and vector level is what I wanted. The big things I wanted to check were
a) The masking predicated loads/stores
and b) The order of the loads/stores to shared memory is preserved after vector distribution

Both are mostly tested in other places, but good to have a longer test showing everything working nicely. WRT the embedded lowering config, I think I'll leave it as is for now and move it later if we end up using fancier logic for dynamic reductions.